linera-core/src/client/mod.rs

```diff
@@ -2331,6 +2341,76 @@ impl<Env: Environment> ChainClient<Env> {
         Ok(())
     }
 
+    /// Downloads only the specific sender blocks needed for missing cross-chain messages.
+    /// This is a targeted alternative to `find_received_certificates` that only downloads
+    /// the exact sender blocks we're missing, rather than searching through all received
+    /// certificates.
+    #[instrument(level = "trace")]
+    async fn download_missing_sender_blocks(
+        &self,
+        missing_blocks: BTreeMap<ChainId, Vec<BlockHeight>>,
+    ) -> Result<(), ChainClientError> {
+        if missing_blocks.is_empty() {
+            return Ok(());
+        }
+
+        let (_, committee) = self.admin_committee().await?;
+        let nodes = self.client.make_nodes(&committee)?;
+        let (max_epoch, committees) = self.client.admin_committees().await?;
+
+        // Download certificates for each sender chain at the specific heights.
+        let certificates = stream::iter(missing_blocks.into_iter())
```

Comment:
IIUC, the issue was here – correct me if I'm wrong @bart-linera  – that this `certificates` collection was slow to download.